src/pip/_internal/index.py

```diff
@@ -73,6 +85,87 @@ def _get_content_type(url, session):
     return resp.headers.get("Content-Type", "")
 
 
+class _SchemeIsVCS(Exception):
+    pass
+
+
+class _NotHTML(Exception):
+    pass
+
+
+def _get_html_response(url, filename, session):
+    """Access an HTML page with GET, and return the response.
+
+    This consists of five parts:
+
+    1. Check whether the scheme is supported (non-VCS). Raises `_SchemeIsVCS`
+       on failure.
+    2. Check whether the URL looks like an archive. If it does, make a HEAD
+       request to check the Content-Type header is indeed HTML, and raise
+       `_NotHTML` if it's not. Raise HTTP exceptions on network failures.
+    3. If URL scheme is file: and points to a directory, make it point to
+       index.html instead.
+    4. Actually perform the request. Raise HTTP exceptions on network failures.
+    5. Check whether Content-Type header to make sure the thing we got is HTML,
+       and raise `_NotHTML` if it's not.
+    """
+    # Check for VCS schemes that do not support lookup as web pages.
+    vcs_scheme = _match_vcs_scheme(url)
+    if vcs_scheme:
+        raise _SchemeIsVCS(vcs_scheme)
+
+    for bad_ext in ARCHIVE_EXTENSIONS:
```

Comment:
How about breaking out a helper function that simply returns whether the file name ends in a bad extension? Then you can have a simple `if` in your function instead of a for loop, which should make the calling function a lot flatter and more readable.